### PR TITLE
Auto select ids for relay

### PR DESCRIPTION
**Pull request by tbezman**
## Description

Relay automatically selects the `id` field on a type if one is available, event when the engineer did not explicitly select it in their query / fragment / whatever. At Gallery, we use graphql-codegen to generate type safe mocks for our tests. This leads to a divergence between the two technologies. The types generated by graphql-codegen do not have the `id` field in the types. This leads to Relay freaking out when the `id` field is not present. 

This PR simply automatically selects the field if one is present.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

I added a new test block to `ts-document-spec` with some good coverage related to this code.

**Test Environment**:

- OS: MacOS
- `@graphql-codegen/...`: ?
- NodeJS: 16.12.0

## Checklist:

- [ ] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by tbezman** [comment]
@charlypoly I'm a bit confused on how I can help move this PR forward. Should we be discussing in the other thread? https://github.com/dotansimha/graphql-code-generator/issues/7888

**Comment by Robinnnnn** [comment]
@charlypoly can we get this merged?

File: packages/plugins/typescript/operations/tests/ts-documents.spec.ts [hunk 1]
```diff
@@ -6441,4 +6441,105 @@ function test(q: GetEntityBrandDataQuery): void {
       `);
     });
   });
+
+  describe.only('autoSelectIds', () => {
```
Comment (n1ru4l) [inline packages/plugins/typescript/operations/tests/ts-documents.spec.ts:6445]: please run all tests

File: packages/plugins/typescript/operations/src/visitor.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ export interface TypeScriptDocumentsParsedConfig extends ParsedDocumentsConfig {
   immutableTypes: boolean;
   noExport: boolean;
   maybeValue: string;
+  autoSelectId: boolean;
```
Comment (n1ru4l) [inline packages/plugins/typescript/operations/src/visitor.ts:27]: Should this also allow providing a list of potential id field names and ID scalars?

**Review by n1ru4l — CHANGES_REQUESTED** [review]
Hey @bezman, it would help if you could update the pull request template as well as address the failing tests. I am not against introducing this!

**Comment by tbezman** [comment]
@n1ru4l definitely. i've since fixed a few small bugs. will get this PR updated asap.

File: packages/plugins/typescript/operations/tests/ts-documents.spec.ts (hunk 1 above)
Comment (tbezman) [reply packages/plugins/typescript/operations/tests/ts-documents.spec.ts:6445]: 🤦

**Comment by tbezman** [comment]
@n1ru4l Alright, we should be good for a review

File: packages/plugins/typescript/operations/src/visitor.ts (hunk 2 above)
Comment (tbezman) [reply packages/plugins/typescript/operations/src/visitor.ts:27]: Are there any frameworks other than Relay that rely on this behavior? I'd be happy to implement if we can come up with some reasonable use cases.

**Comment by n1ru4l** [comment]
Hey @tbezman, starting with https://github.com/dotansimha/graphql-code-generator/releases/tag/release-1676995873 it is now possible to provide custom document transforms.

This allows implementing this in user-land, therefore this PR might no longer be needed.

What do you think?

**Comment by n1ru4l** [comment]
Hey @tbezman any update regarding https://github.com/dotansimha/graphql-code-generator/pull/7881#issuecomment-1443393864?

**Pull request closed without merging**